### PR TITLE
[SPARK-27750][CORE] Standalone scheduler - ability to prioritize applications over drivers, many drivers act like Denial of Service

### PR DESCRIPTION
**Pull request by tooptoop4**
### What changes were proposed in this pull request?
Don't allow drivers to use all the cores in spark standalone scheduler. Reserve some cores for apps

### Why are the changes needed?
Airflow is triggering many Spark jobs in parallel, spark standalone gets 'full' with drivers and can never execute apps as drivers took all the cores. The spark cluster essentially becomes stuck and never completes any drivers (as no cores avail to run the associated apps), never frees up cores. Manual intervention is needed to either kill some drivers (but airflow can retry sending them), or force scale up of the spark cluster by registering more workers (than ASG MaxSize).

### Does this PR introduce any user-facing change?
By default no, they must opt-in by explicitly setting config value > 0

### How was this patch tested?
Running in prod for 3 months

### CODE REVIEW

**Review by Ngone51 — COMMENTED** [review]
How about limiting the number of concurrent running drivers? Personally, I feel it's more convenient for user to set # of driver rather than the cores. WDYT?

**Comment by tooptoop4** [comment]
@Ngone51 but each driver can have different amount of cores

**Comment by Ngone51** [comment]
> @Ngone51 but each driver can have different amount of cores

Yeah, but for such case, many drivers race for the reserved core concurrently, how can we know which drivers win the race and decide the amount of cores for them?

**Comment by tooptoop4** [comment]
@Ngone51 as long as some cores are not used by any driver, then i don't mind which drivers got in first. I just want some cores that can only be consumed by apps

**Comment by Ngone51** [comment]
> I just want some cores that can only be consumed by apps

When we limit the number of running drivers, we'll definitely have reserved cores for the apps.

**Comment by tooptoop4** [comment]
@Ngone51 that is worse, one day you could have 10 drivers each taking 1 core, another day you could have 5 drivers each taking 4 cores (in this case if u limited on 10 drivers, all cores would be used). So it does not work well with diff sized drivers. Cores is simple/uniform way

**Comment by jiangxb1987** [comment]
What's blocking you from choosing the client deploy mode? Using the client deploy mode would let you launch the driver program locally and thus no need to worry about drivers involved into  cluster resource contentions.

**Comment by jiangxb1987** [comment]
Speaking of the proposed improvement, it does somehow migrates the issue by allowing you to launch at least a few executors, but it would probably still be the case that not enough cluster resources are going to executors, thus many drivers still need to wait for free slots to launch their pending tasks. Without other considerations, I think you should give client deploy mode a try (or you already have other issues thus not able to choose the client mode) ?

**Comment by tooptoop4** [comment]
@jiangxb1987 client deploy mode is subject to limits of single master machine. if I want to run 200+ spark-submits in parallel then the master machine must have a enough memory to support all those drivers. With my change few executors is all I need, but even I can reserve 100s of cores for just apps with this new config

**Comment by jiangxb1987** [comment]
Sorry I don't have enough context to understand your use case, but submitting 200+ applications at the same time to a Spark cluster is something I'm not expecting. Basically I would expect a lot less applications, and each application can submit a few jobs, thus we don't really need to launch that much of drivers.

**Comment by tooptoop4** [comment]
i am doing a data lake, where 10000s of files of different schemas get ingested daily. what is the reluctance to merging this?

**Comment by tooptoop4** [comment]
can u pls merge @dongjoon-hyun ?

File: core/src/main/scala/org/apache/spark/internal/config/Deploy.scala [hunk 1]
```diff
@@ -75,5 +75,10 @@ private[spark] object Deploy {
     .intConf
     .createWithDefault(Int.MaxValue)
 
+  val CORES_RESERVED_FOR_APPS = ConfigBuilder("spark.deploy.coresReservedForApps")
+    .version("2.4.6")
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/internal/config/Deploy.scala]: Hi, @tooptoop4 . We don't backport a feature. 
The feasible next version is 3.1.0.

**Comment by jiangxb1987** [comment]
The proposed improvement is just not necessary. I don't see any need to submit 200+ applications, you could start one application and submit multiple jobs.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 2]
```diff
@@ -797,28 +798,32 @@ private[deploy] class Master(
...
-        numWorkersVisited += 1
-        if (canLaunchDriver(worker, driver.desc)) {
-          val allocated = worker.acquireResources(driver.desc.resourceReqs)
-          driver.withResources(allocated)
-          launchDriver(worker, driver)
-          waitingDrivers -= driver
-          launched = true
+    val allFreeCores = shuffledAliveWorkers.map(_.coresFree).sum
+    val forDriversFreeCores = math.max(allFreeCores - coresReservedForApps, 0)
+    if (forDriversFreeCores > 0) {
```
Comment (dongjoon-hyun) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Could you make a test case for this?

**Comment by tooptoop4** [comment]
@jiangxb1987 ur suggestion is don't scale?

**Comment by jiangxb1987** [comment]
What do you mean by scale? You don't need 200 drivers, you can still launch one application and submit your jobs, this way your workload should work.

**Comment by tooptoop4** [comment]
1. one app is going to be slower 2. what if 7 out of 200 job fails, whole app fails meaning poor resumability. 3. files arrive at different times, so can't wait for all to arrive and then do single app. This PR fixes clear bug, spark standalone gets itself stuck. if you don't like the number 200, think of submitting 6 apps in cluster mode on a 4 core cluster, spark gets full of drivers and no apps can ever run, spark stays stuck forever. With this fix, I can guarantee some cores for apps so spark never gets stuck with only drivers

**Comment by jiangxb1987** [comment]
I‘m -1 to this change, because it's trying to resolve an issue that doesn't even exist. 
Please read https://spark.apache.org/docs/latest/cluster-overview.html before you ask. Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you for your proposal, @tooptoop4. However, according to the above discussion, I'll close this PR. Thank you, @tooptoop4 , @Ngone51 , @jiangxb1987 .

**Pull request closed without merging**